cmd/apikey.go

```diff
@@ -0,0 +1,593 @@
+	printText("Expiration ..... %s", humanTime(apikey.ExpirationMs))
+	printText("Id ............. %s", apikey.ID)
+	printText("API Key ........ %s (won't be shown again)", apikey.Key)
+	printText(`Credentials .... %s (use it as "Authorization: APIKey <credentials>" header to communicate with APM Server, won't be shown again)`,
+		credentials)
+
+	printJSON(struct {
+		es.CreateAPIKeyResponse
+		Privileges es.CreatePrivilegesResponse `json:"created_privileges,omitempty"`
+	}{
+		CreateAPIKeyResponse: apikey,
+		Privileges:           privilegesCreated,
+	})
+}
+
+func getAPIKey(client es.Client, id, name *string, validOnly, asJSON bool) {
+	if isSet(id) {
+		name = nil
+	} else if isSet(name) {
+		id = nil
+	}
+	request := es.GetAPIKeyRequest{
+		APIKeyQuery: es.APIKeyQuery{
+			ID:   id,
+			Name: name,
+		},
+	}
+
+	apikeys, err := es.GetAPIKeys(client, request)
+	if err != nil {
+		printErr(err, asJSON)
+		return
+	}
+
+	transform := es.GetAPIKeyResponse{APIKeys: make([]es.APIKeyResponse, 0)}
+	printText, printJSON := printers(asJSON)
+	for _, apikey := range apikeys.APIKeys {
+		expiry := humanTime(apikey.ExpirationMs)
+		if validOnly && (apikey.Invalidated || expiry == "expired") {
+			continue
+		}
+		creation := time.Unix(apikey.Creation/1000, 0).Format("2006-02-01 15:04")
+		printText("Username ....... %s", apikey.Username)
+		printText("Api Key Name ... %s", apikey.Name)
+		printText("Id ............. %s", apikey.ID)
+		printText("Creation ....... %s", creation)
+		printText("Invalidated .... %t", apikey.Invalidated)
+		if !apikey.Invalidated {
+			printText("Expiration ..... %s", expiry)
+		}
+		printText("")
+		transform.APIKeys = append(transform.APIKeys, apikey)
+	}
+	printText("%d API Keys found", len(transform.APIKeys))
+	printJSON(transform)
+}
+
+func invalidateAPIKey(client es.Client, id, name *string, deletePrivileges, asJSON bool) {
+	if isSet(id) {
+		name = nil
+	} else if isSet(name) {
+		id = nil
+	}
+	invalidateKeysRequest := es.InvalidateAPIKeyRequest{
+		APIKeyQuery: es.APIKeyQuery{
+			ID:   id,
+			Name: name,
+		},
+	}
+
+	invalidation, err := es.InvalidateAPIKey(client, invalidateKeysRequest)
+	if err != nil {
+		printErr(err, asJSON)
+		return
+	}
+	printText, printJSON := printers(asJSON)
+	out := struct {
+		es.InvalidateAPIKeyResponse
+		Privileges []es.DeletePrivilegeResponse `json:"deleted_privileges,omitempty"`
+	}{
+		InvalidateAPIKeyResponse: invalidation,
+		Privileges:               make([]es.DeletePrivilegeResponse, 0),
+	}
+	printText("Invalidated keys ... %s", strings.Join(invalidation.Invalidated, ", "))
+	printText("Error count ........ %d", invalidation.ErrorCount)
+
+	for _, privilege := range auth.PrivilegesAll {
+		if !deletePrivileges {
+			break
+		}
+		deletePrivilegesRequest := es.DeletePrivilegeRequest{
+			Application: auth.Application,
+			Privilege:   privilege.Name,
+		}
+
+		deletion, err := es.DeletePrivileges(client, deletePrivilegesRequest)
+		if err != nil {
+			continue
```

Comment:
IIRC, because it would fail to delete privileges if they don't exist. But I did this more than 1 month ago, my memory is fuzzy. Should check again

Reply:
Ah, because we're returning an error on 404. Probably shouldn't do that, but treat it as an expected status code. A problem for another day.

Reply:
isn't that what this it is doing? or you mean inside `DeletePrivileges` impl? what difference would that make? :thinking:

Reply:
I meant inside DeletePrivleges - we could also do it here, if we expose details of the error. This isn't checking for 404, it's ignoring _all_ errors. The error could be because Elasticsearch is down, because the privilege doesn't exist, or because it's a Monday.